Update qwen3_coder_detector.py for streaming

maocheng23 opened the pull request:
## Motivation

This PR is to improve the streaming tool call parser from reading all values to incremental flow. 

## Modifications

Update the tool call parser for qwen3_coder_detector using XML format. 
- **Streaming Function Names**: Tool names are now sent immediately upon detection with empty parameters
- **Incremental Parameter Streaming**: Parameters are streamed as JSON diffs when each XML parameter block completes
- **State Management**: Added proper streaming state tracking variables and reset mechanisms
- **Error Handling**: Fixed `AttributeError` issues that caused connection drops during streaming 

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

Checked with local test
<img width="568" height="220" alt="6e08bf6889faac0da9de32bfdf96156" src="https://github.com/user-attachments/assets/87e69ce6-1fa0-4491-a5f0-570cac8d82e8" />

maocheng23 committed:
Update qwen3_coder_detector.py for streaming

Update qwen3_coder_detector with streaming

maocheng23 committed:
Merge branch 'main' into maocheng23-patch-1

CatherineSue commented on python/sglang/srt/function_call/qwen3_coder_detector.py:
```diff
@@ -70,23 +84,168 @@ def parse_streaming_increment(
+                        
+                        self.streamed_args_for_tool[self.current_tool_id] += argument_diff
+                    
+                    self._current_parameters = new_params
+                    self.prev_tool_call_arr[self.current_tool_id]["arguments"] = new_params
+                
+                # Check if tool call is complete
+                if self.tool_call_end_token in self._buf:
+                    end_pos = self._buf.find(self.tool_call_end_token)
+                    
+                    # Process any remaining complete parameters before the end token
+                    before_end = self._buf[:end_pos]
+                    final_matches = list(re.finditer(r'<parameter=([^>]+)>(.*?)</parameter>', before_end, re.DOTALL))
+                    
+                    final_params = {}
+                    for match in final_matches:
+                        param_name = match.group(1).strip()
+                        param_value = match.group(2)
+                        final_params[param_name] = _safe_val(param_value)
+                    
+                    if final_params != self._current_parameters:
+                        final_args_json = json.dumps(final_params, ensure_ascii=False)
+                        sent_length = len(self.streamed_args_for_tool[self.current_tool_id])
+                        
+                        if len(final_args_json) > sent_length:
+                            final_diff = final_args_json[sent_length:]
+                            calls.append(ToolCallItem(
+                                tool_index=self.current_tool_id,
+                                name=None,
+                                parameters=final_diff,
+                            ))
+                    
+                    # Complete the tool call
+                    self._buf = self._buf[end_pos + len(self.tool_call_end_token):]
+                    self._reset_streaming_state()
+                    self.current_tool_id += 1
+                    continue
+                else:
+                    # Tool call not complete yet, wait for more text
+                    break
```
I agree a helper function might be better.

CatherineSue reviewed (commented):
Thanks for all the improvements. Overall logic is sound to me. Left some comments for details.

maocheng23 committed:
Merge branch 'sgl-project:main' into maocheng23-patch-1

maocheng23 commented on python/sglang/srt/function_call/qwen3_coder_detector.py:
```diff
@@ -70,23 +84,168 @@ def parse_streaming_increment(
-                break
-            block = self._buf[: e + len(self.tool_call_end_token)]
-            self._buf = self._buf[e + len(self.tool_call_end_token) :]
-            calls.extend(self._parse_block(block, tools))
```
It's still be used in detect_and_parse function

maocheng23 committed:
local commit to address comments, but still need test

maocheng23 committed:
Merge branch 'sgl-project:main' into maocheng23-patch-1

maocheng23 commented:
Hi @CatherineSue , d702004 addressed your comments and local test passed. Please let me know if this looks good to you. Thanks a lot!
<img width="1138" height="301" alt="image" src="https://github.com/user-attachments/assets/fe74c1fb-cabb-4e4f-bccf-e080c79b3638" />

CatherineSue commented on python/sglang/srt/function_call/qwen3_coder_detector.py:
```diff
@@ -56,6 +57,19 @@ def __init__(self):
+        self._function_name_sent: bool = False
+
+    def _get_tool_indices(self, tools: List[Tool]) -> Dict[str, int]:
+        """Get a mapping of tool names to their indices in the tools list."""
```
Seems like this function is still here?

maocheng23 committed:
fix the logic and passed the unit-test

maocheng23 committed:
Merge branch 'maocheng23-patch-1' of https://github.com/maocheng23/sglang into maocheng23-patch-1

CatherineSue commented on python/sglang/srt/function_call/qwen3_coder_detector.py:
```diff
@@ -70,23 +83,207 @@ def parse_streaming_increment(
+                    json_fragment = f'{{{json.dumps(key, ensure_ascii=False)}: {json.dumps(value, ensure_ascii=False)}'
+                else:
+                    # Multiple parameters in first batch - send complete JSON
+                    json_fragment = json.dumps(new_params, ensure_ascii=False)
```
Will this be problematic? This will make json_fragment complete. For instance, if this new_params is `{"a": 2, "b": 3}`

Then later at line 252, we have `{"c": 5}`?

maocheng23 committed:
address comments and pass tests

maocheng23 commented:
Unit test passed with local server and commented the TestOpenAIPythonicFunctionCalling 

<img width="1222" height="222" alt="image" src="https://github.com/user-attachments/assets/9d1bfefb-5344-448a-9b8a-350baf9a4ac6" />

maocheng23 committed:
Merge branch 'sgl-project:main' into maocheng23-patch-1

CatherineSue committed:
Merge branch 'main' into maocheng23-patch-1

maocheng23 committed:
Merge branch 'sgl-project:main' into maocheng23-patch-1

maocheng23 committed:
fix CI issues and update ut

maocheng23 committed:
Merge branch 'sgl-project:main' into maocheng23-patch-1

CatherineSue committed:
Merge branch 'main' into maocheng23-patch-1

CatherineSue committed:
Merge branch 'main' into maocheng23-patch-1

CatherineSue committed:
Merge branch 'main' into maocheng23-patch-1

CatherineSue committed:
Merge branch 'main' into maocheng23-patch-1
